mailpile/plugins/vcard_face.py

```diff
@@ -0,0 +1,57 @@
+import time
+
+from mailpile.i18n import gettext as _
+from mailpile.mailutils import Email
+from mailpile.plugins import PluginManager
+from mailpile.search import MailIndex
+from mailpile.vcard import MailpileVCard, VCardImporter, VCardLine
+
+
+_plugins = PluginManager(builtin=__file__)
+
+
+class FaceImporter(VCardImporter):
+    """
+    Get avatar from Face header.
+    """
+    FORMAT_NAME = 'Face'
+    FORMAT_DESCRIPTION = _('Get avatar from Face header.')
+    SHORT_NAME = 'face'
+    CONFIG_RULES = {
+        'active': [_('Enable this importer'), bool, True],
+    }
+    VCARD_TS = 'x-face-ts'
+    VCARD_IMG = ''
+
+    def get_vcards(self):
+        mail_index = MailIndex(self.session.config)
+        mail_index.load(self.session)
+        results = []
+
+        for vcard in self.session.config.vcards.values():
+            # Get latest message from vcard email address.
+            from_messages = list(mail_index.search(
+                self.session, ['from:' + vcard.email]).as_set())
+            if not from_messages:
+                continue
+            mail_index.sort_results(self.session, from_messages, 'date')
+            latest_message = Email(mail_index, from_messages.pop()).get_msg()
```

Comment:
> make the search more exact, so it doesn't return any results if messages from that user don't have a Face header.

This would be nice regardless of other performance optimizations, but I didn't find an existing way to do this. Would this require adding search infrastructure?

> don't perform the search every time if the user already has an avatar (maybe use a random number and only search 1/1000 times, or see if there is a timestamp somewhere we can check and use)

It seems like the expense of this operation compared with other vcard lookups that make remote calls is relative to the machine -- is it constrained for local resources (cpu/memory) or bandwidth? Wouldn't the same caching optimizations would be relevant to all of them?